clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -2985,92 +2985,79 @@ static void handleWeakImportAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
-          (getExprValue(AL.getArgAsExpr(1), S.getASTContext()) <=
-           getExprValue(A->getYDim(), S.getASTContext())) &&
-          (getExprValue(AL.getArgAsExpr(2), S.getASTContext()) <=
-           getExprValue(A->getZDim(), S.getASTContext())))) {
-      S.Diag(AL.getLoc(), diag::err_conflicting_sycl_function_attributes)
-          << AL << A->getSpelling();
-      Result &= false;
-    }
-  }
-
-  if (const auto *A = D->getAttr<ReqdWorkGroupSizeAttr>()) {
-    if (!((getExprValue(AL.getArgAsExpr(0), S.getASTContext()) >=
-           getExprValue(A->getXDim(), S.getASTContext())) &&
-          (getExprValue(AL.getArgAsExpr(1), S.getASTContext()) >=
-           getExprValue(A->getYDim(), S.getASTContext())) &&
-          (getExprValue(AL.getArgAsExpr(2), S.getASTContext()) >=
-           getExprValue(A->getZDim(), S.getASTContext())))) {
-      S.Diag(AL.getLoc(), diag::err_conflicting_sycl_function_attributes)
-          << AL << A->getSpelling();
-      Result &= false;
-    }
-  }
-  return Result;
+  if (AL.getKind() == ParsedAttr::AT_SYCLIntelMaxGlobalWorkDim) {
+    // In case the value of 'max_global_work_dim' attribute equals to 0
+    // we shall ensure that if max_work_group_size and reqd_work_group_size
+    // attributes exist, they hold equal values (1, 1, 1).
+    if (const auto *A = D->getAttr<SYCLIntelMaxWorkGroupSizeAttr>()) {
+      if (getExprValue(A->getXDim(), S.getASTContext()) != 1 ||
+          getExprValue(A->getYDim(), S.getASTContext()) != 1 ||
+          getExprValue(A->getZDim(), S.getASTContext()) != 1) {
+        S.Diag(A->getLocation(), diag::err_sycl_x_y_z_arguments_must_be_one)
+            << A << AL;
+      }
+    }
+    if (const auto *A = D->getAttr<ReqdWorkGroupSizeAttr>()) {
+      if (getExprValue(A->getXDim(), S.getASTContext()) != 1 ||
+          getExprValue(A->getYDim(), S.getASTContext()) != 1 ||
+          getExprValue(A->getZDim(), S.getASTContext()) != 1) {
+        S.Diag(A->getLocation(), diag::err_sycl_x_y_z_arguments_must_be_one)
+            << A << AL;
+      }
+    }
+  } else {
+    // in other cases we should check that attribute value
+    // >= reqd_work_group_size attribute value and
+    // <= max_work_group_size attribute value
+    if (const auto *A = D->getAttr<SYCLIntelMaxGlobalWorkDimAttr>()) {
+      int64_t AttrValue =
+          A->getValue()->getIntegerConstantExpr(S.Context)->getSExtValue();
+      if (AttrValue == 0) {
+        if (!((getExprValue(AL.getArgAsExpr(0), S.getASTContext()) == 1) &&
+              (getExprValue(AL.getArgAsExpr(1), S.getASTContext()) == 1) &&
+              (getExprValue(AL.getArgAsExpr(2), S.getASTContext()) == 1))) {
+          S.Diag(AL.getLoc(), diag::err_sycl_x_y_z_arguments_must_be_one)
+              << AL << A;
+        }
+      }
+    } else if (const auto *A = D->getAttr<SYCLIntelMaxWorkGroupSizeAttr>()) {
+      if (!((getExprValue(AL.getArgAsExpr(0), S.getASTContext()) <=
```

Comment:
I am starting to wonder if `ParsedAttr` should get an `arg_as_expr_begin()`, `arg_as_expr_end()`, and `args_as_exprs()` functions so that we can access the argument expressions as an array (and use it for things like the range-based APIs). I don't know if there would be struggles with implementing it, but I feel like it has the potential to clean up a lot of this redundancy.

(I don't insist on the change, but if you'd like to give it a shot to see if you like the results better, I also won't complain.)